New feature: Automatic heatbed cooldown fan control

soimy opened the pull request:
Add pin definition HEATBED_AUTO_FAN_PIN in configuration_adv.h
When defined, the firmware will automatically monitor the difference between current_heatbed_temperature and target_heatbed_temperature. 
If the difference is huge ( Cooling-down ), the heat bed cooling fan will spin to speed up the cooldown process. 

This is especially useful when PEI build platform is used and users wish to peel-down the printed part quicker.

soimy committed:
New feature: Automatic heatbed cooldown fan control

Add pin definition HEATBED_AUTO_FAN_PIN in configuration_adv.h
When defined, the firmware will automatically monitor the difference between current_heatbed_temperature and target_heatbed_temperature.
If the difference is huge ( Cooling-down ), the heat bed cooling fan will spin to speed up the cooldown process.

This is especially useful when PEI build platform is used and users wish to peel-down the printed part quicker.

soimy commented:
Tested on my UM2-clone for about a week, just work.

FATALHERMIT commented:
@soimy  are you sure you uploaded the right version? because the autonomous check system can't build it for a hephestos 2 because the variables weren't declared on the right scope , here's the log : 
```
The command "use_example_configs Hephestos_2" exited with 0.
8.94s$ build_marlin
Picked up JAVA_TOOL_OPTIONS: 
Loading configuration...
Xlib:  extension "RANDR" missing on display ":1.0".
Initializing packages...
Preparing boards...
Verifying...
sketch/Marlin_main.cpp: In function 'bool shove_and_echo_command(const char*, bool)':
sketch/Marlin_main.cpp:825:21: warning: invalid conversion from 'const char*' to 'char*' [-fpermissive]
     echo_command(cmd);
                     ^
sketch/Marlin_main.cpp:799:13: note: initializing argument 1 of 'void echo_command(char*)'
 inline void echo_command(char * const cmd) {
             ^
sketch/Marlin_main.cpp: In function 'bool enqueue_and_echo_command(const char*, bool)':
sketch/Marlin_main.cpp:905:21: warning: invalid conversion from 'const char*' to 'char*' [-fpermissive]
     echo_command(cmd);
                     ^
sketch/Marlin_main.cpp:799:13: note: initializing argument 1 of 'void echo_command(char*)'
 inline void echo_command(char * const cmd) {
             ^
sketch/temperature.cpp: In static member function 'static void Temperature::checkHeatbedAutoFan()':
temperature.cpp:491: error: 'HEATBED_AUTO_FAN_PIN' was not declared in this scope
     const int8_t fanPin = HEATBED_AUTO_FAN_PIN;
                           ^
temperature.cpp:497: error: 'HEATBED_AUTO_FAN_DELTA_TEMPERATURE' was not declared in this scope
     if (delta_temperature_bed > HEATBED_AUTO_FAN_DELTA_TEMPERATURE &&
                                 ^
temperature.cpp:498: error: 'HEATBED_AUTO_FAN_MIN_TEMPERATURE' was not declared in this scope
         (int)current_temperature_bed > HEATBED_AUTO_FAN_MIN_TEMPERATURE ) {
                                        ^
temperature.cpp:501: error: 'HEATBED_AUTO_FAN_SPEED' was not declared in this scope
         digitalWrite(fanPin, HEATBED_AUTO_FAN_SPEED);
                              ^
exit status 1
'HEATBED_AUTO_FAN_PIN' was not declared in this scope
The command "build_marlin" exited with 1.
0.02s$ restore_configs
```

Bob-the-Kuhn commented:
Congratulations on doing your first PR!!!

----

Please don't put the HEATBED_AUTO_FAN_PIN in configuration_adv.h.  

The preferred location for pin definitions is in the pins_YOUR_BOARD.h files.

Yes, you can find other pin definitions in the configuration files but those are not good examples to follow.

The nasty part is you should also update the other pins_YOUR_BOARD.h files.

And another gotcha is you should update the file pinsDebug.h.  It's used when doing debugging via the M43 command.  It has a list of the known pin names in it.  You'll need to update that list.

Bob-the-Kuhn commented:
Here's something I expect you'll eventually need to consult.  It's a guide on how to update a PR, resolve the failed checks and squash all the commits into one.  

It's in the process of being added to the Marlin Wiki so feedback is requested.

[updating the PR.zip](https://github.com/MarlinFirmware/Marlin/files/869550/updating.the.PR.zip)

soimy commented:
Thanks for the direction. I will update the code after testing on my printer asap.

soimy commented:
Fix all compile error and add Pindebug.h , squash commit as the doc's direction. Thanks a lot. This is my first PR to marlin community 👏

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -224,6 +224,16 @@
+// 
+// When the delta temperature between current and target bed temperature is big, 
+// heatbed cooldown fan will automaticly blow to accelerate the cooldown proccess.
+// Useful when you have a PEI build surface.
```
The comment is long enough to use a block comment. With some tweaks…
```cpp
/**
 * Heatbed Cooldown Fan
 * 
 * When the target bed temperature is below the current temperature by the
 * set delta value, the fan will automatically turn on to cool the bed.
 * Useful when you have a PEI build surface.
 */
```

thinkyhead commented on Marlin/example_configurations/Cartesio/Configuration_adv.h:
```diff
@@ -224,6 +224,16 @@
+// When the delta temperature between current and target bed temperature is big, 
+// heatbed cooldown fan will automaticly blow to accelerate the cooldown proccess.
+// Useful when you have a PEI build surface.
+#define HEATBED_AUTO_FAN_PIN -1
```
As mentioned, for completeness, you can define the pin that will be used as the default `HEATBED_AUTO_FAN_PIN` in each `pins_SOMEBOARD.h` file. For `HEATBED_AUTO_FAN_PIN` this is easy to figure out. The pin will be same as `FAN1_PIN` or `FAN2_PIN`. So in `pins_RAMBO.h`, the block would be:
```cpp
#ifndef HEATBED_AUTO_FAN_PIN 
  #define HEATBED_AUTO_FAN_PIN 2
#endif
```
…which corresponds to its `FAN2_PIN`. Just using `FAN2_PIN` should also work in most cases (but maybe not RAMPS-based). For most, you should probably set it to `FAN1_PIN` or some other known-to-be-free 12V MOSFET pin.

The `pins_MEGACONTROLLER.h` file has `CONTROLLERFAN_PIN` defined as the same as `FAN2_PIN`, and that particular conflict isn't handled in `Conditionals_post.h`, so that (line 451) would need to be tweaked for that case.

soimy replied:
Quite confused on this part. I was not able to access so many kinds of motherboards but I think most 3d printers on the market don't feature a heat bed cooling fan, so this is more likely a DIY custom feature (_Not associated with any specific motherboard_). 

On my own MKS Gen1.4 board I need to change predefined 24V E1 MOSFET pin to FAN1_PIN to enable extra fan. If I want even more fan control I need sth. like RRD fan extender or [LED pwm control board](https://www.aliexpress.com/item/XD-93A-3W-LED-PWM-dimming-driver-support-IN-7-30V-OUT-700mA-5pcs-lot-free/32298096364.html) alike and wire any unused pwm pin to pwm signal input. Then it's not a motherboard business. 

I think editing `HEATBED_AUTO_FAN_PIN` in `configuration_adv.h` is more straight forward ( Just like `CONTROLLERFAN_PIN & E_AUTO_FAN_PIN` ).

thinkyhead replied:
> I think editing HEATBED_AUTO_FAN_PIN in configuration_adv.h is more straight forward  

That is of course retained. But in addition to the options in `Configuration_adv.h`, the pins files provide the most likely pin for each accessory. In this case, the pin will have to be a 12V pin, and thus one of the MOSFET pins.

You can see how the `En_AUTO_FAN_PINS` are handled, centrally, in the main `pins.h` file:
```cpp
//
// Assign auto fan pins if needed
//
#if !defined(E0_AUTO_FAN_PIN) && defined(ORIG_E0_AUTO_FAN_PIN)
  #define E0_AUTO_FAN_PIN ORIG_E0_AUTO_FAN_PIN
#endif
#if !defined(E1_AUTO_FAN_PIN) && defined(ORIG_E1_AUTO_FAN_PIN)
  #define E1_AUTO_FAN_PIN ORIG_E1_AUTO_FAN_PIN
#endif
#if !defined(E2_AUTO_FAN_PIN) && defined(ORIG_E2_AUTO_FAN_PIN)
  #define E2_AUTO_FAN_PIN ORIG_E2_AUTO_FAN_PIN
#endif
#if !defined(E3_AUTO_FAN_PIN) && defined(ORIG_E3_AUTO_FAN_PIN)
  #define E3_AUTO_FAN_PIN ORIG_E3_AUTO_FAN_PIN
#endif
```
This style was chosen to avoid having to put blocks like this in several board pins files:
```cpp
#ifndef E0_AUTO_FAN_PIN
  #define E0_AUTO_FAN_PIN 5
#endif
#ifndef E1_AUTO_FAN_PIN
  #define E1_AUTO_FAN_PIN 5
#endif
#ifndef E2_AUTO_FAN_PIN
  #define E2_AUTO_FAN_PIN 5
#endif
#ifndef E3_AUTO_FAN_PIN
  #define E3_AUTO_FAN_PIN 5
#endif
```
…and instead they can just contain…
```cpp
#define ORIG_E0_AUTO_FAN_PIN 5
#define ORIG_E1_AUTO_FAN_PIN 5
#define ORIG_E2_AUTO_FAN_PIN 5
#define ORIG_E3_AUTO_FAN_PIN 5
```
Either style allows you to:
- turn off the pin in `Configuration_adv.h` by setting it to -1 (or empty),
- override the pin's default value in `Configuration_adv.h` by setting it to some value, or
- use the default pin set in the pins file by commenting out the define in `Configuration_adv.h`.

soimy replied:
Got it, I will add default heat bed pin for boards which have unused FANn_PIN

> turn off the pin in Configuration_adv.h by setting it to -1 (or empty),
> override the pin's default value in Configuration_adv.h by setting it to some value, or
> use the default pin set in the pins file by commenting out the define in Configuration_adv.h.

The action _commenting out to use default pin for my board_ looks a bit weird to me. Shall we do it in this order: 

1. Leave the line commented in `Configuration_adv.h` to use the default pin defined by `pin_SOMEBOARD.h`;
2. `pin.h` will handle the turning-off (set to -1) if pin was not defined in `pin_SOMEBOARD.h`;
3. User can either disable or override that pin manually in `Configuration_adv.h`.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
 
+  void Temperature::checkHeatbedAutoFan() {
+    const int8_t fanPin = HEATBED_AUTO_FAN_PIN;
+    if(fanPin <= 0) return; // Early exit with no heatbed fan defined
```
One space after `if`.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
+    int delta_temperature_bed = (int)current_temperature_bed - target_temperature_bed;
+
+    if (delta_temperature_bed > HEATBED_AUTO_FAN_DELTA_TEMPERATURE &&
+        (int)current_temperature_bed > HEATBED_AUTO_FAN_MIN_TEMPERATURE ) {
```
Remove extra space before `)`.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
+
+    if (delta_temperature_bed > HEATBED_AUTO_FAN_DELTA_TEMPERATURE &&
+        (int)current_temperature_bed > HEATBED_AUTO_FAN_MIN_TEMPERATURE ) {
+      if(!fanState){
```
Add one space after `if` and one space before `{`.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
+        digitalWrite(fanPin, HEATBED_AUTO_FAN_SPEED);
+        analogWrite(fanPin, HEATBED_AUTO_FAN_SPEED);
+      }
+    } else {
```
Place `else` on a new line:
```cpp
  }
  else {
    . . .
  }
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
+        analogWrite(fanPin, HEATBED_AUTO_FAN_SPEED);
+      }
+    } else {
+      if(fanState) {
```
Add one space after `if`.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,30 @@ int Temperature::getHeaterPower(int heater) {
+    if(fanPin <= 0) return; // Early exit with no heatbed fan defined
+
+    static bool fanState = false;
+    int delta_temperature_bed = (int)current_temperature_bed - target_temperature_bed;
```
If not changing, use `const`.
```cpp
const int delta_temperature_bed = (int)current_temperature_bed - target_temperature_bed;
```

thinkyhead reviewed (commented):
A few style tweaks are needed, plus the default pins. Overall, looking pretty good!

Roxy-3D commented:
@soimy Don't feel bad...   He dislikes my indentation and capitalization much worse than yours!

soimy commented:
Thanks to all of you guiding me on my first PR @thinkyhead @Bob-the-Kuhn @Roxy-3D ! 
I'm more of a designer than a coder and has a lot to learn.

soimy committed:
Code styling fix of Heatbed auto fan codes

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -224,6 +224,18 @@
+ * When the target bed temperature is below the current temperature by the
+ * set delta value, the fan will automatically turn on to cool the bed.
+ * Useful when you have a PEI build surface.
+ */
```
One file down, 17 to go!

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -487,6 +487,31 @@ int Temperature::getHeaterPower(int heater) {
 
+  void Temperature::checkHeatbedAutoFan() {
+    const int8_t fanPin = HEATBED_AUTO_FAN_PIN;
+    if( fanPin <= 0) return; // Early exit with no heatbed fan defined
```
Please see https://github.com/MarlinFirmware/Marlin/wiki/DNE-Coding-Standards for code formatting guidelines.

soimy committed:
Add board default pins
and some code styling fix

soimy commented:
[Added] default heat bed pin for boards which have unused FANn_PIN. Please check if they are properly set @thinkyhead 

> turn off the pin in Configuration_adv.h by setting it to -1 (or empty),
> override the pin's default value in Configuration_adv.h by setting it to some value, or
> use the default pin set in the pins file by commenting out the define in Configuration_adv.h.

The action _commenting out to use default pin for my board_ looks a bit weird to me. Shall we do it in this order? : 

1. Leave the line commented in `Configuration_adv.h` to use the default pin defined by `pin_SOMEBOARD.h`;
2. `pin.h` will handle the turning-off (set to -1) if pin was not defined in `pin_SOMEBOARD.h`;
3. User can either disable or override that pin manually in `Configuration_adv.h`.

soimy committed:
Merge branch 'RCBugFix' into Headbed-auto-fan

soimy committed:
Add support for new board example config

Also fix in new pinDebug_list.h

soimy committed:
Fix conditional bugs for auto fans

After add support for 5th extruder, some conditions in conditianal_post.h is missing and now be fixed

soimy commented on Marlin/Conditionals_post.h:
```diff
@@ -452,7 +452,8 @@
   #define HAS_AUTO_FAN_4 (HOTENDS > 4 && PIN_EXISTS(E4_AUTO_FAN))
-  #define HAS_AUTO_FAN (HAS_AUTO_FAN_0 || HAS_AUTO_FAN_1 || HAS_AUTO_FAN_2 || HAS_AUTO_FAN_3)
+  #define HAS_HEATBED_AUTO_FAN (PIN_EXISTS(HEATBED_AUTO_FAN))
+  #define HAS_AUTO_FAN (HAS_AUTO_FAN_0 || HAS_AUTO_FAN_1 || HAS_AUTO_FAN_2 || HAS_AUTO_FAN_3 || HAS_AUTO_FAN_4 || HAS_HEATBED_AUTO_FAN)
```
@thinkyhead When 5th extruder auto fan pin is now available, this condition should also update.

soimy commented on Marlin/Conditionals_post.h:
```diff
@@ -464,9 +465,9 @@
-  #define HAS_FAN2 (PIN_EXISTS(FAN2) && CONTROLLERFAN_PIN != FAN2_PIN && E0_AUTO_FAN_PIN != FAN2_PIN && E1_AUTO_FAN_PIN != FAN2_PIN && E2_AUTO_FAN_PIN != FAN2_PIN && E3_AUTO_FAN_PIN != FAN2_PIN)
-  #define HAS_CONTROLLERFAN (PIN_EXISTS(CONTROLLERFAN))
+  #define HAS_FAN1 (PIN_EXISTS(FAN1) && CONTROLLERFAN_PIN != FAN1_PIN && E0_AUTO_FAN_PIN != FAN1_PIN && E1_AUTO_FAN_PIN != FAN1_PIN && E2_AUTO_FAN_PIN != FAN1_PIN && E3_AUTO_FAN_PIN != FAN1_PIN && E4_AUTO_FAN_PIN != FAN1_PIN && HEATBED_AUTO_FAN_PIN != FAN1_PIN)
+  #define HAS_FAN2 (PIN_EXISTS(FAN2) && CONTROLLERFAN_PIN != FAN2_PIN && E0_AUTO_FAN_PIN != FAN2_PIN && E1_AUTO_FAN_PIN != FAN2_PIN && E2_AUTO_FAN_PIN != FAN2_PIN && E3_AUTO_FAN_PIN != FAN2_PIN && E4_AUTO_FAN_PIN != FAN2_PIN && HEATBED_AUTO_FAN_PIN != FAN2_PIN)
```
And these two lines too.

fiveangle commented:
cool ! 😉

rjacobs1969 commented:
I hope this get merged into master soon

FredrikJ1 commented:
I need this. It takes a long time for the heatbed to cooldown so the print can be removed. I dont want to remove the glass. Good work.

fiveangle commented:
Was this feature implemented ?
